### PR TITLE
Fix model preview

### PR DESCRIPTION
**Pull request by haslinghuis**
Fixes fast spinning of model preview in receiver tab.

### CODE REVIEW

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> FAIL
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> FAIL
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

**Comment by Asizon** [comment]
Wow good work @haslinghuis looks so good

**Comment by McGiverGim** [comment]
@haslinghuis what was the problem exactly? The PR has no description. I only see that we copy the code from rates to here.
If we change all this code, all of them duplicate from there, maybe it's time to think about adding it to the RateCurve library.

**Comment by haslinghuis** [comment]
Yes as I couldn't find the real bug I started duplicating the working code from pid_tuning. Moved the initialization of currentRates to RatesCurve.

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> FAIL
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

**Review by blckmn — DISMISSED** [review]

**Comment by hydra** [comment]
this fixed it for me, but I feel it is important to actually understand what the real issue in `master` is rather than blindly duplicating code.

**Comment by haslinghuis** [comment]
Problem was in RateCurve library using pidtuning as reference. Decoupled now as receiver is using same code.
Thanks to @mrRobot62 for debugging.

File: src/js/fc.js
Comment (McGiverGim) [inline src/js/fc.js:940]: Adding UI with one tab in the `fc.js` ? This is not the way to go.

**Review by McGiverGim — CHANGES_REQUESTED** [review]

File: src/js/fc.js
Comment (haslinghuis) [reply src/js/fc.js]: Good find, moved this part back to pid_tuning tab.

**Comment by mrRobot62** [comment]
Tested this fix several times on my Mac. Works now as designed.
No fast spinning copter on start.

Perfect - good job

**Review by limonspb — APPROVED** [review]

**Comment by limonspb** [comment]
wondering why it shows 13% code duplication.

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> PASS
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

**Review by limonspb — APPROVED** [review]

File: src/js/RateCurve.js
Comment (McGiverGim) [inline src/js/RateCurve.js:168]: I'm not too sure about this change, it implies that we now use the FC rates and not the ones in the TAB? The idea is to show the model preview and can change it live in the TAB. Are we modifying the FC value while the user changes the TAB values?
In either way, I think the best way is to pass an object to the method with the values that we must to represent. Use the FC makes this library dependent on the FC file and is not necessary.
Maybe the `currentRates` object that returns the new method?

File: src/js/fc.js
Comment (McGiverGim) [inline src/js/fc.js:906]: To me is difficult to see this object here. In my opinion all this conversions must be done either:
- When reading the FC values using MSP directly. But maybe this complicates the code more.
- Or using an utility method, in the `RateCurve.js` library.

**Review by McGiverGim — COMMENTED** [review]
Some doubts about the code...

**Comment by haslinghuis** [comment]
@limonspb can't find the duplicated block

image

File: src/js/RateCurve.js
Comment (haslinghuis) [reply src/js/RateCurve.js:243]: Fixed here (as we use this in two tabs:

image

Comment (haslinghuis) [reply src/js/RateCurve.js:243]: It's also fixes changing rate type and updating currentRates,

**Review by limonspb — APPROVED** [review]

**Review by limonspb — APPROVED** [review]

**Review by ctzsnooze — APPROVED** [review]

**Review by limonspb — CHANGES_REQUESTED** [review]
plz squash :)

**Review by ctzsnooze — APPROVED** [review]

**Review by Asizon — APPROVED** [review]
Looks so good!

**Review by McGiverGim — APPROVED** [review]
I continue thinking that is a bad idea to use the FC object directly inside the RateCurve.js library, but ok, I will approve it.

**Review by chmelevskij — APPROVED** [review]

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> PASS
 - assigned to an approver -> PASS
 - approver count at least three -> PASS

**Review by blckmn — APPROVED** [review]

**Comment by blckmn** [comment]
SonarCloud code smell?

**Pull request merged**
